### DISCUSSION TITLE
Maybe use semantic versioning and semantic release?

### DISCUSSION THREAD

**Discussion by buehler** [FilledStacks/stacked#887] [Ideas]
Hey @ferrarafer and @FilledStacks 

I don't know which versioning strategy you use in this repository, I just wanted to pitch an idea. In my other repositories, I use semantic versioning in combination with semantic-release. Do you know about the system and/or have you considered it?

As far as I know, it does not work well with mono-repos, but it does a pretty nice job at updating versioning numbers, changelogs, and the likes.

As an example: https://github.com/smartive/zitadel-dart
When the repository is updated, semantic release will determine the new version, update the pubspec and the changelog according to the commit message and then pushes the latest version to pub.dev. All fully automatically. You could even do pre-releases if configured.

Wdyt?

**Comment by FilledStacks** [comment]
@buehler that's a great idea, I haven't used it but the more we can automate the happier I would be. I'm very open to it. 

We haven't had the time to solidify our pipeline and automate this process. Because we're currently only building stacked based on our requirements it's always a code -> immediate release type thing, if the CI passes. 

With that said this would be a welcomed feature. I'm currently forking the AutoRoute package into our generator so we can have a Navigator2 implementation for users to choose from which turns out to have been a lot more work than expected. My time is a little locked up on that now in preparation for the new Flutter Web Series. 

I think having that setup and me hardening some of the tests we can definitely do that. 

One question is how does it choose the version number? I don't know much about the system.

**Comment by buehler** [comment]
Hey @FilledStacks

Happy to explain :-)

[Semantic Versioning](https://semver.org/) clearly states what needs to happen for the version number to change. They also support preview / release candidate / whatever you name it - releases (e.g. v1.0.0-prerelease.1). Most packaging systems adhere to this standard since it makes updating pretty easy. All non breaking changes (i.e. a major version increase) must be compatible.

That said, semantic release makes use of that and removes the "human factor" in choosing an updated version number. I encountered the flaw myself when I needed to release a new version and was thinking about "is it really this or that?". Semantic release uses the [Angular Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard to determine the next version number for the release. The three main parts of determining the new numbers are:
- `fix ...` -> create a new patch version
- `feat ...` -> create a new minor version
- Any commit with `BREAKING CHANGE: ...` in it -> create a new major version

Examples:
- `fix(cli): fixed a bug`
- `fix: fixed a bug`
- `feat(generator): added error handling`
- ```
  fix: made something un-nullable

  BREAKING CHANGE: this breaks the function and
  people need to migrate by using this and that.
  ```

The rest of the format can be reviewed in the aforementioned link to the conventional commits.

We've been using this system for several years now and it makes releasing such a breeze. Adhering to this format not only allows you to select version numbers, but also automatically generates changelogs and release notes. You can review this in almost all of our [smartive](https://github.com/smartive) repositories.

That said, a possible setup for stacked could be:

Since mono-repos are not particularly well supported, split all the packages into separate repositories.

This allows [renovate bot](https://github.com/renovatebot/renovate) (or dependabot...) to update dependencies in your packages one by one. So, if something in the core changes, all dependent packages get updated after the release as well.

Use semantic release in all the sub packages, and if needed create a "master" repo that links to all other packages with git submodules (to have all the packages in one place).

With that, all packages can be updated/developed on their own and are released when there are dependency-changes (renovate bot), or fixes/features/breaking changes. Nearly everything can be automated since you already have a premium stock of tests.

The only two things you'd need for the automation would be:
- A personal access token of the user that deploys the package
- The pub.dev credentials to actually deploy the package

I just created another automatically released package yesterday: https://github.com/smartive/flutter-generic-dropdown-widget which automatically releases with semantic release.

I'd be happy to help you with the setup if you require assistance ;-)

I hope I explained it in an understandable way

**Reply by FilledStacks** [reply]
This is a really great explanation and makes a ton of sense. I appreciate you going into the details like that. I'll think about using this, my only concern is with the separate repo's. Not specifically because I like mono repos but because I'm a single maintainer (in general) for these packages and the mental capacity that 11 repo's would take up is a bit too much for me. 

Having everything in one place for issues, etc is very helpful. But I really do like the versioning automation. 

I also assume that for cases where there's no breaking changes but you've added many features in prep for a major version bump you can still trigger the major version bump and the automation process. Like our Stacked v3 release had no breaking changes but a lot of new functionality.

**Reply by buehler** [reply]
I can totally understand the concern :-) In my opinion, it is not a big issue. Several big frameworks do it like this (also this is what I'd recommend): Create an organization for the framework (to ease the management of contributors and such), create a repository for issues, create the repositories for the packages / addons / plugins / whatnot and deactivate issues in them. Such that issues and discussions are only in the central repository.

For the other topic about the major version increase: You _can_, but you _shouldn't_. That's the primary reason why semantic releasing eliminates the "Human Factor" in versioning. Since you don't break anything, you don't need a major version increase. It is only a subjective matter that it must be "a major version" since there are new features. All people that have automatic dependency update will automatically benefit from new features if they are not in a new major version since they (normally) get updated automatically.
Versioning is just a number. You don't need a new major version to state: "see the new major stacked framework version". This works with 3.1 as well as it does with 2.10. But in case you really want to, just make a fake breaking change somewhere ;)

There are two possibilities (which I've both been using):
- Create a new version on each commit to the default branch
- Use a next branch to create pre-releases for each merge (PR) and then periodically or manually merge the next branch into the default branch to create the "real" next version

**Reply by FilledStacks** [reply]
@buehler you definitely have way more git knowledge than me 🙌 and this change will make it easier for others to contribute and get their work merged in. 

I think its a good step up for the repo and the project and would be happy with the change. I would really appreciate your help with this. I'll create the Stacked organisation on github. If it's not too much to ask, do you think you can find the time to do the entire setup as you mentioned?

There's no rush (since I didn't even think about this haha), I simply have no idea where to start. And then another favor would be, if you're done and happy with the setup if we could get on a call and you take me through the setup. Whenever it's ready. I'd like to get a better understanding of it.

**Reply by buehler** [reply]
Hey @FilledStacks
Sure, happy to help :-) I do love frameworks like this and I'm excited to help you as much as I can. It surely will help managing further contributions / contributors and everything else.

And of course, we can discuss the solution when I'm finished setting it up :-)

You'll just need to create the organization and add me as an organization admin (username "buehler"). Then I should be able to set everything up. When I'm finished, you'll need to provide two repository (organization) secrets: a personal access token from github, that is able to access the repositories (I'll guide you in the call) and the pub.dev credentials to publish the packages.

**Reply by buehler** [reply]
by the way: if you want, I can also create the org, it does not really matter :-)
I'll start as soon as you created the org or have told me that I should do it ;-)

Have a nice evening and/or weekend :-)

**Reply by FilledStacks** [reply]
@buehler thank you very much! I have setup the Stacked-Org and invited you. I appreciate your time and effort you're about to put in.

**Reply by buehler** [reply]
Perfect, just accepted the invite :-)
Can you make me an owner? (At least for the duration of the setup)
Otherwise I cannot setup some parts.

**Reply by FilledStacks** [reply]
Done! 🚀

**Reply by buehler** [reply]
Alright, so far so good.
I created and configured all the necessary basics. (The org does not use any pro features, so I downgraded to free again, it is all open source).

Now I'm going to move the code from the original stacked repo to the org repos and setup the tags and stuff for automatic deployment.

When this is done, the last step would be to create/update the PAT and the pub dev credentials and move this repository to the org under the name "framework". Reason being that you don't lose any stars and/or issues.

Then this repository (new "framework") will be repurposed to the "master" repository that contains submodules to all repositories in the org and is mainly used for development purposes and issue / discussion management. Effective merging and stuff is done in the repositories.

**Reply by FilledStacks** [reply]
This is awesome @buehler I received many notifications over the weekend so I knew you were making a lot of progress. This is really cool! We can go through the required updates later on our call. I'd like to keep the name stacked for the repo if possible instead of changing it to framework.

**Comment by buehler** [comment]
as far as I know, there are also github actions which check if the commit message is adhering to conventional commits.

**Comment by buehler** [comment]
Just came to my mind: another version of the automatic deployment, if you don't what to split the repository: leave all packages in one big mono-repo, but then, when something gets merged, _ALL_ packages get deployed with the new version (e.g. https://github.com/buehler/dotnet-operator-sdk). So, all packages stick to the same versioning.

There are pro's and con's to both versions. Personally I'd stick with the separated repos and the one master repo that uses submodules for development / testing the interaction between the packages.

**Reply by FilledStacks** [reply]
@buehler the repo split is one of my concerns. But we can probably still do everything as it is now using a master repo and the sub modules. It's quite a lot of work 😅 I don't know if the effort will out weigh the reward at this stage. 

Could also be that I just finished a massive amount of navigator 2 work in preparation for the new Flutter Web series and just want to start building the damn Flutter web app haha. 

What do you think is the time / effort involved in setting this up?

**Reply by buehler** [reply]
The reason for the master repo would just be local development. From there you could create branches in each repo and develop a specific feature in one specific repo. In conjunction with my reply above: you could use this "master" repo to gather issues and discussions together.

And I totally understand the thing about navigator2 and your flutter app ;-)

I think splitting and setting up everything would take maybe, half a day?
As I said, I'd happy to help you if you need assistance :-)

As mentioned above, I would setup an organization, create all repos in there, create the master/issues repo with all packages as submodules to ease issue/discussion/development management, then setup renovate bot for all of the repositories and move the current packages in their own respective repos with automated testing/releasing in place.

**Comment by FilledStacks** [comment]
@buehler the cli package is not auto publishing the version. 

When I look at the release logs it doesn't pick up that there are required changes. The commits start with feat or fix but it doesn't pick it up. I'll do a manual release for now because I need the version live to complete my tutorial.

**Reply by buehler** [reply]
Wait :)
Do you have a colon after the topic?

**Reply by buehler** [reply]
It MUST be :
fix: foobar
Or
feat: foobar

Not just
Fix foobar

**Reply by buehler** [reply]
If you do a manual, it will confuse the system afterwards. 
You can do a manual, but then you need to create a tag in the repo as well (correctly)

**Reply by buehler** [reply]
Just looked at it, your commit message is "update web template"

That is not a recognized commit.

You need to use
"feat: update web template"

**Reply by FilledStacks** [reply]
Yes it has feat: 

```
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: Update web template (#4)

* feat: Update Flutter App intro animation to fadeIn only

* feat: Update web templates to use CustomRoutes

* Corrects broken main template for web
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The commit should not trigger a release
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: chore(deps): update dependency lints to v2 (#1)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The commit should not trigger a release
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: Updates web templates (#3)

* feat: Adds updated splash loader to the html

* feat: Adds flutter into animation to avoid jump from splash when running web app
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The commit should not trigger a release
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: Beta-merge (#2)

* feat: merge from beta branch

* Adding unsaved files

* chore: apply dart format fix

* Remove example folder

* chore: fix test.yml workflow to run build_runner before analyze
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The commit should not trigger a release
[5:36:52 AM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analysis of 4 commits complete: no release
[5:36:52 AM] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[5:36:52 AM] [semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/exec"
[5:36:52 AM] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/exec"
[5:36:52 AM] [semantic-release] › ℹ  There are no relevant changes, so no new version is released.
```

**Reply by buehler** [reply]
No it has in the commit description, but not the first line. The first line of the commit is what counts

**Reply by buehler** [reply]
You can create an empty commit to create a deployment
"Git commit --allow-empty -m "feat: whatever"

**Reply by FilledStacks** [reply]
I need to start my recording process I'm already behind 😄 I'll use a local installation of my package for now and start my recording, I'll only be online again in about 6 hours after all recording is done. 

If you can spot the issue let me know. I have no idea why it's not picking it up.

**Reply by FilledStacks** [reply]
Mmmmmmm, those are all first line only commits. 

It might have been the squash and merge. I've never used that either, I only rebase. Will try quickly with an small change

**Reply by FilledStacks** [reply]
It worked @buehler it was probably when the squash and merge happened when I enabled the auto-merge afterwards.

This. means not all my changes show up in the change log. How do I avoid that?

**Reply by buehler** [reply]
Make one PR for one feat/fix. or: make one commit that "covers" multiple things. But best practice is when you actually make one PR per thing.

**Reply by FilledStacks** [reply]
@buehler that's gonna slow me down quite a bit. Stacked is still in rapid development phase sometimes I need to add 10+ changes into the cli to continue with my project or the next step of a tutorial. 

I'll have to figure out something soon, speed is super important to me. I'll make changes to the workflows to reduce some of the overhead if it gets in the way too much. I already have a refined workflow to produce the open source work + content  I create on a weekly basis, adding this kind of PR overhead will most likely increase the time for implementation and release by an hour+ which in most cases is 5x the time. I usually do multiple fixes and push within 15-30 minutes of listing all the fixes needed for the next step. 

BUT, we'll see. I'll give it another week and track the time spent on overhead.

**Reply by buehler** [reply]
Ofcourse you can just stay in main/master and just commit your stuff there. Then you can go for a commit per feature. Then you have everything in the changelog when you push it.

**Reply by buehler** [reply]
then you'll be able to do that ;)
create 3 commits like:
- feat: new
- fix: bug 1
- fix: bug 2

and create these in your main branch, then everything will get released at once (when you push)

**Reply by FilledStacks** [reply]
Interesting...  but then we're taking out the PR workflow completely. Which is also fine I guess.

I think re-base might help with this. I'll allow rebased on the branches and see. 

This will replay all the commits on main without duplication and should pick up the same as you mentioned above even when commiting on a different branch.

**Reply by buehler** [reply]
Yea, rebase & merge can also help. Then you just need to pay attention to the "proper" committing rules. All commits, that somehow (by any means) end up in the main branch and adhere to the angular commit format are picked up. We usually use squash & merge because we use the one branch for one topic strategy. How you manage the repository is totally up to you :)

**Reply by FilledStacks** [reply]
Awesome. I think I'll do rebase so the change log per version can have more than one change.
